p4c-ubpf: new uBPF back-end for p4c

osinstom opened the pull request:
This PR implements the uBPF (Userspace BPF) backend for the P4 compiler

The `p4c-ubpf` compiler allows to translate P4 programs into the uBPF programs.

The backend for uBPF is mostly based on p4c-ebpf, but generates the C code, which is compatible with the user space BPF implementation. Moreover, it has support for P4 registers and programmable actions (packet's modifications + tunneling).

We tested `p4c-ubpf` with the Oko switch [1], but the uBPF Virtual Machine can be used in any software switch implementing the kernel bypass (e.g. DPDK apps).

Basic usage:

```bash
# compile P4 program to C code
$ p4c-ubpf -o test.c PROGRAM.p4 
# compile test.c to BPF machine code
$ clang-6.0 -O2 -I .. -target bpf -c test.c -o /tmp/test.o
```

[1] https://github.com/Orange-OpenSource/oko/tree/p4rt-ovs

mbudiu-vmw commented:
I took a quick look at the PR, it looks pretty good.
I think that the main thing missing is integrating your tests in the existing test suite. 
The tests can be run at several levels: just compiling - and this we should always to - and running the programs on generated traffic. Since you depend on a lot of infrastructure these tests may be disabled by default - like the ebpf kernel tests are today.

I don't know if we can setup Travis to run Vagrant, DPDK and all the other stuff necessary for your tests. If we can, we should add your tests to the continuous integration, otherwise it's very easy for them to get broken by some random commit elsewhere.

For just the compiler tests we could add your files to testdata/p4_16_samples, like we do with the ebpf tests (I am assuming they can be compiled with p4test); we could also suffix all your test programs with something like -ubpf.p4, which would make it easy for the test scripts to pick up only your tests for running them. 

We can leave your test programs where they are as well, but then it would still be nice for `make check` to run `p4c-ubpf` on your programs at the very least.

I will start reviewing your contribution, but it may take a while. Perhaps I will submit multiple reviews so you can start addressing the issues one by one.

mbudiu-vmw commented:
I also assume that you are committing to maintaining this implementation and fixing bugs that people may report for the foreseeable future.

osinstom commented:
Ok. We will work to integrate our tests with existing test suite. We'll take a look at how it is done for ebpf and, then, prepare a separate commit. 

> I also assume that you are committing to maintaining this implementation and fixing bugs that people may report for the foreseeable future.

Definitely! We devote some time to maintaining `p4c-ubpf`.

jnfoster commented:
Just to be clear, I think what Mihai is proposing is to split the testing into:
- Compile-only tests, which can be run whether or not the backend is installed
- uBPF-specific tests, which would run if the backend were installed

Cheers,
-N

On Sat, Jan 11, 2020 at 10:16 AM Tomek Osiński <notifications@github.com<mailto:notifications@github.com>> wrote:

Ok. We will work to integrate our tests with existing test suite. We'll take a look at how it is done for ebpf and, then, prepare a separate commit.

I also assume that you are committing to maintaining this implementation and fixing bugs that people may report for the foreseeable future.

Definitely! We devote some time to maintaining p4c-ubpf.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/p4lang/p4c/pull/2134?email_source=notifications&email_token=AAOCFUQNTRZC23SQNEGTWILQ5HPF7A5CNFSM4KFJABPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIWEBBY#issuecomment-573325447>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAOCFUQ7DTQ5DUXV56OWVV3Q5HPF7ANCNFSM4KFJABPA>.

mbudiu-vmw commented on backends/ubpf/examples/simple-firewall.p4:
```diff
@@ -0,0 +1,181 @@
+    apply {
+
+        if (hdr.tcp.isValid()) {
+            @atomic {
```
you have implemented support for atomic?
(haven't got to that part of the code yet)

osinstom replied:
Removed.

mbudiu-vmw commented on backends/ubpf/p4c-ubpf.cpp:
```diff
@@ -0,0 +1,94 @@
+    auto hook = options.getDebugHook();
+    bool isv1 = options.langVersion == CompilerOptions::FrontendVersion::P4_14;
+    if (isv1) {
+        ::error("This compiler only handles P4-16");
```
you should consider migrating to the function that takes an `ErrorType::ERR_*` argument, but there is no hurry.

osinstom replied:
Done.

mbudiu-vmw commented on backends/ubpf/p4c-ubpf.cpp:
```diff
@@ -0,0 +1,94 @@
+    EBPF::MidEnd midend;
+    midend.addDebugHook(hook);
+    auto toplevel = midend.run(options, program);
+    if (options.dumpJsonFile)
```
what is the json for?

osinstom replied:
Removed. It was useless.

mbudiu-vmw commented on backends/ubpf/p4include/ubpf_model.p4:
```diff
@@ -0,0 +1,50 @@
+                filter<H, M> filt,
+                deparser<H> dprs);
+
+extern void mark_to_drop();
```
I would suggest documenting these functions.

jafingerhut replied:
In particular, for mark_to_drop() and mark_to_pass(), what is the default behavior after a packet is done processing?  Probably pass?

Then if the P4 program does mark_to_drop() and/or mark_to_pass() calls, whichever one was last determines whether the packet is passed or dropped later?

And there are no user-visible P4 variables that are modified by mark_to_drop() or mark_to_pass()?  I would not ask, except this was an issue for several years with v1model's mark_to_drop(), and why it was changed to take a parameter later.  It appears you have nothing that corresponds to standard_metadata, so it would make sense if in your architecture, mark_to_drop() and mark_to_pass() only modify state "hidden from the user's P4 program" inside the architecture implementation.

Another good thing to document about these extern functions is which of the architecture's parsers/controls they are allowed to be called in, e.g. maybe you intended something like "mark_to_drop() and mark_to_pass() may only be called in the filter control".

Comments in the p4include/v1model.p4 file would be good to look at for examples.  If any of them happen to match the behavior you provide (e.g. that might be true for your Register extern compared to v1model's register extern), feel free to copy and paste whatever is accurate for your architecture.

mbudiu-vmw commented on backends/ubpf/tests/environment/setup-pktgen.sh:
```diff
@@ -0,0 +1,67 @@
+if [ ! -d "/home/vagrant/p4c" ]
+then
+ cd /home/vagrant
+ git clone https://github.com/P4-Research/p4c.git
```
you may need to change this after the PR is accepted. You should add a TODO somewhere.

mbudiu-vmw replied:
It appears in several other places.

osinstom replied:
Done.

mbudiu-vmw commented on backends/ubpf/tests/environment/setup-switch.sh:
```diff
@@ -0,0 +1,183 @@
+# Install clang-10
+if ! type "/vagrant/llvm-project/build/bin/clang" > /dev/null; then
+  cd /vagrant
+  git clone https://github.com/P4-Research/llvm-project
```
so you have your own clone of llvm as well?
it would be nice to use an off-the-shelf version

osinstom replied:
Yep, we are aware of that.

This is related to the limitation of BPF itself. It was also discussed in regards of p4c-xdp [1]. As we are in the userspace, we can simply increase the BPF's stack size. However, it requires to have a new flag in `clang` compiler (this is also on our TODO list), as proposed in this thread. 

Currently, we have our fork of `clang` / llvm, in which the stack size is increased and hardcoded.

[1] https://github.com/vmware/p4c-xdp/issues/22
[2] http://lists.llvm.org/pipermail/cfe-users/2017-March/001126.html

mbudiu-vmw commented on backends/ubpf/tests/testdata/oko-test-actions.p4:
```diff
@@ -0,0 +1,167 @@
+/*
```
what is the difference between the program in examples and this one?

mbudiu-vmw replied:
if they are the same then it would be nice to avoid the duplication. symlinks?

osinstom replied:
We've added symlinks.

mbudiu-vmw commented on backends/ubpf/ubpfControl.cpp:
```diff
@@ -0,0 +1,772 @@
+        BUG("%1%: not yet handled", decl);
+    }
+
+    bool UBPFControl::determineIfVariableIsPointer(const IR::Declaration_Variable *vd, const EBPF::EBPFType *etype) {
```
Wouldn't it be cleaner to have a separate Inspector pass run before your code generation and build a map of all variables that should be used as pointers?

osinstom replied:
That's good idea. I've just refactored the code to build vector of pointer variables before emitting them.

mbudiu-vmw commented on backends/ubpf/ubpfControl.cpp:
```diff
@@ -0,0 +1,772 @@
+    void UBPFControlBodyTranslator::processFunction(
+            const P4::ExternFunction *function) {
+        if (function->method->name.name == control->program->model.drop.name) {
+            builder->append("pass = false");
```
I think you want to use `passVariable` here instead of `pass`.

osinstom replied:
Done.

mbudiu-vmw commented on backends/ubpf/ubpfControl.h:
```diff
@@ -0,0 +1,137 @@
+        explicit UBPFControlBodyTranslator(const UBPFControl *control);
+
+        virtual void adjustPacketHead(const IR::Expression *expression, bool add);
+
```
These empty lines are not very useful.

osinstom replied:
Done.

mbudiu-vmw commented on backends/ubpf/ubpfDeparser.cpp:
```diff
@@ -0,0 +1,235 @@
+                                                          unsigned alignment,
+                                                          EBPF::EBPFType *type) {
+
+        unsigned widthToEmit = dynamic_cast<EBPF::IHasWidth *>(type)->widthInBits();
```
Do you want to check whether the dynamic_cast can return NULL?

osinstom replied:
I've added additional check.

mbudiu-vmw commented on backends/ubpf/ubpfDeparser.cpp:
```diff
@@ -0,0 +1,235 @@
+            loadSize = 32;
+        } else if (widthToEmit <= 64) {
+            swap = "htonll";
+            loadSize = 64;
```
The ebpf compiler actually has a bug in the handling of htons; it is not clear that fields that do not have a size a power of 2 should be transformed using these macros. For example, the ebpf compiler incorrectly handles ethernet addresses. But perhaps you have tested this stuff. It is not very clear what header fields should be converted to machine order; I think that the ideal case would be not convert to machine order anything on parsing, but only when a field is used in some arithmetic operation.

osinstom replied:
That's certainly a place for further optimizations. In comparison to ebpf compiler we have fixed the issue with Ethernet address (48 bit, < 64), IPv6 addresses (128 bit, > 64) and also the problem with writing a custom-width fields using `write_partial`, which didn't work properly in `ebpf` compiler.

mbudiu-vmw replied:
That's great. If you could at some point send a PR to fix that back-end it would be great.

mbudiu-vmw commented on backends/ubpf/ubpfDeparser.cpp:
```diff
@@ -0,0 +1,235 @@
+    bool UBPFDeparser::build() {
+        auto pl = controlBlock->container->type->applyParams;
+        if (pl->size() != 2) {
+            ::error("Expected deparser to have exactly 2 parameters");
```
You should supply some IR::Node argument to error.

osinstom replied:
Done.

mbudiu-vmw commented on backends/ubpf/ubpfDeparser.h:
```diff
@@ -0,0 +1,81 @@
+        virtual void compileEmit(const IR::Vector<IR::Argument> *args);
+
+        bool preorder(const IR::MethodCallExpression *expression) override;
+
```
Maybe you can remove some of these empty lines.

osinstom replied:
Done.

mbudiu-vmw commented on backends/ubpf/ubpfParser.cpp:
```diff
@@ -0,0 +1,361 @@
+                helper = "load_word";
+                loadSize = 32;
+            } else {
+                if (widthToExtract > 64) BUG("Unexpected width %d", widthToExtract);
```
This seems to be impossible since you just checked that widthToExtract <= 64 above.

osinstom replied:
I removed this check. It is useless, indeed.

mbudiu-vmw commented on backends/ubpf/ubpfParser.cpp:
```diff
@@ -0,0 +1,361 @@
+        unsigned width = ht->width_bits();
+        auto program = state->parser->program;
+        builder->emitIndent();
+        // TODO: Check if packet is not too short
```
This should not be too difficult... can you please do this?

osinstom replied:
That TODO is not outdated (to be removed). 

The packet length's check is implemented in [1].

[1] https://github.com/p4lang/p4c/blob/9b6868a0d0eb48d3b2dced4e43ada1aa20636e9b/backends/ubpf/ubpfProgram.cpp#L278

mbudiu-vmw commented on backends/ubpf/ubpfTable.h:
```diff
@@ -0,0 +1,91 @@
+        void emitTypes(EBPF::CodeBuilder *builder);
+
+        void emitActionArguments(EBPF::CodeBuilder *builder,
+                                 const IR::P4Action *action, cstring name);
```
fewer empty lines?

osinstom replied:
Done.

mbudiu-vmw commented on backends/ubpf/ubpfTable.cpp:
```diff
@@ -0,0 +1,408 @@
+                auto ef = mi->to<P4::ExternFunction>();
+                if (ef != nullptr) {
+                    if (ef->method->name.name == program->model.drop.name) {
+                        builder->append("pass = false");
```
passVariable?

osinstom replied:
Done.

mbudiu-vmw commented on backends/ubpf/ubpfTable.cpp:
```diff
@@ -0,0 +1,408 @@
+
+            void convertActionBody(const IR::Vector<IR::StatOrDecl> *body) {
+                for (auto s : *body) {
+                    if (!s->is<IR::Statement>()) {
```
usually you just need to iterate visit over all statements and write a method for each statement kind.
that's the point of the visitor pattern.
If you will support if statements in actions then this code will be wrong.

osinstom replied:
I refactored a little bit this method, so that it visits all the statements now. We based the implementation of this method based on the `convertActionBody()` from `action.cpp` (simple_switch).

mbudiu-vmw commented on backends/ubpf/ubpfTable.cpp:
```diff
@@ -0,0 +1,408 @@
+                property->value->is<IR::ExpressionValue>()) {
+                auto pExpressionValue = property->value->to<IR::ExpressionValue>();
+                auto pConstant = pExpressionValue->expression->to<IR::Constant>();
+                this->size = pConstant->asInt();
```
what if this is larger than UINT16_MAX?

osinstom replied:
Refactored.

mbudiu-vmw commented on backends/ubpf/ubpfTable.cpp:
```diff
@@ -0,0 +1,408 @@
+        size = UINT16_MAX; // Default value 2^16. Next power is too big for ubpf vm.
+                           // For instance, 2^17 causes error while loading program.
+        for (auto property : properties) {
+            if (property->name.name == table->container->properties->sizePropertyName &&
```
Maybe you can use the `getConstantProperty` method of P4::Table

osinstom replied:
Refactored.

mbudiu-vmw commented on backends/ubpf/ubpfRegister.cpp:
```diff
@@ -0,0 +1,174 @@
+        this->valueType = valueType;
+        this->keyType = keyType;
+
+        if (valueType->is<IR::Type_Name>()) {
```
Probably the right way to do this is to look it up in the TypeMap.

osinstom replied:
Could you elaborate more on that? I'm not sure I got the point.

mbudiu-vmw replied:
This should have been a bit higher: instead of using `di->type` you should do `typeMap->getType(di)`.

osinstom replied:
Done.

mbudiu-vmw commented on backends/ubpf/ubpfRegister.cpp:
```diff
@@ -0,0 +1,174 @@
+                                        const IR::MethodCallExpression *expression) {
+        BUG_CHECK(expression->arguments->size() == 1,
+                  "Expected 1 argument for %1%", expression);
+        auto target = (UbpfTarget *) builder->target;
```
Do you need this cast? Wouldn't virtual dispatch do the right thing anyway?

osinstom replied:
Removed.

mbudiu-vmw commented on backends/ubpf/ubpfProgram.h:
```diff
@@ -0,0 +1,84 @@
+            contextVar = cstring("ctx");
+        }
+
+        bool build() override;
```
can you remove some of these empty lines?

osinstom replied:
I think it's done now.

mbudiu-vmw commented on backends/ubpf/ubpfProgram.cpp:
```diff
@@ -0,0 +1,303 @@
+    }
+
+    void UBPFProgram::emitUbpfHelpers(EBPF::CodeBuilder *builder) const {
+        builder->append(
```
Shouldn't these be in some method of UBPFTarget?

osinstom replied:
Moved to UBPFTarget.

mbudiu-vmw commented on backends/ubpf/ubpfProgram.cpp:
```diff
@@ -0,0 +1,303 @@
+                !d->is<IR::Type_Extern>() && !d->is<IR::Type_Parser>() &&
+                !d->is<IR::Type_Control>() && !d->is<IR::Type_Typedef>() &&
+                !d->is<IR::Type_Error>()) {
+                if (d->toString().startsWith("struct tuple")) {
```
This is weird; what is this supposed to do?

osinstom replied:
Removed. It was useless.

osinstom commented:
@mbudiu-vmw thank you for your comments! I'll address them soon. We also started to prepare compile-only tests. Currently, in parallel, we work on a demo showcase for our internal purposes, so it grabs us a part of time and we cannot work intensively on PR, sorry for that!

osinstom commented on backends/ubpf/README.md:
```diff
@@ -0,0 +1,87 @@
+This repository implements the uBPF (Userspace BPF) backend for the P4 compiler (https://github.com/p4lang/p4c).
+
+The **p4c-ubpf** compiler allows to translate P4 programs into the uBPF programs. We use the uBPF implementation provided 
+by [the Oko/P4rt-OVS switch](https://github.com/Orange-OpenSource/oko/tree/p4rt-ovs). The Oko's uBPF VM is based on the
```
Yep, this is a fork of OVS. I presented the details at OVSCon 19' [1]. There are no official plans to merge it with OVS yet. It is still a research prototype, but once it will be mature enough we can consider to propose such a contribution to OVS. 

However, as far as I know, P4rt-OVS is the only switch to test uBPF back-end.

[1] https://www.openvswitch.org/support/ovscon2019/#s4.3F

osinstom commented on backends/ubpf/docs/EXAMPLES.md:
```diff
@@ -0,0 +1,197 @@
+- `key_type` - is bit array type (i.e. bit<32>) or struct like type  
+- `number_of_elements` - the maximum number of key-value pairs
+
+Currently registers have a limitation - they are not being initialized with default values. Initialization has to be done by a control plane. 
```
We analyzed different design options to initialize BPF maps (P4 registers) before the start of the program. And we decided that intializing them from a control plane is the best option for the moment. We could think about better approach later.

osinstom commented on backends/ubpf/docs/EXAMPLES.md:
```diff
@@ -0,0 +1,197 @@
+
+This is very simple example of stateful firewall. Every TCP packet is analyzed to track the state of the TCP connection. 
+If the traffic belongs to known connection it is passed. Otherwise, it is dropped.  
+Notice that the example program uses hash function which is constrained to hash only 64 bit values - that's why TCP connection is identified via IP source and destination address.  
```
Yep, that's the limitation of our implementation. In particular, it is the limitation of the `ubpf_hash()` helper that we use to compute hash value. Also, to be improved in the future.

osinstom commented:
I've just submitted a small commit (with some minor modifications) addressing some of your comments. In particular, there was some code, which was useless, but remained in the files.

mbudiu-vmw commented on testdata/p4_16_samples_outputs/ipv4-actions_ubpf.c:
```diff
@@ -0,0 +1,358 @@
+/* Automatically generated by p4c-ubpf from /home/mateusz/p4c/testdata/p4_16_samples/ipv4-actions_ubpf.p4 on Mon Jan 20 15:27:07 2020
```
Do you want to save these files in the repository?
Are you comparing with them in some tool? 
The ebpf back-end does not save the C files - I was afraid that they would change too often with compiler changes to make for good tests, but perhaps that's wrong.

osinstom replied:
You're it is not the best way, but I think it is the only reasonable (at least for now) way to test if the C files generated by `p4c-ubpf` are correct. We compare them at very basic level, using a minimalistic Python code.

We are aware that it could require changes to these file once we will add some new functionality to `p4c-ubpf`. I would keep these files for now, until we will integrate P4rt-OVS (or some other uBPF verifier) with `make check-ubpf`.

osinstom replied:
We've removed these files and replaced them by userspace-only tests (based on those implemented for `ebpf`)

osinstom commented on backends/ubpf/README.md:
```diff
@@ -0,0 +1,87 @@
+# Introduction to uBPF Backend
+
+This repository implements the uBPF (Userspace BPF) backend for the P4 compiler (https://github.com/p4lang/p4c).
```
Removed.

osinstom commented on backends/ubpf/examples/gtp.p4:
```diff
@@ -0,0 +1,224 @@
+
+
+struct Headers_t {
+    @name("ethernet")
```
We've removed these annotations. They were useless.

osinstom commented on backends/ubpf/examples/gtp.p4:
```diff
@@ -0,0 +1,224 @@
+limitations under the License.
+*/
+
+#include "ubpf_model.p4"
```
We've fixed that in all P4 files.

osinstom commented on backends/ubpf/ubpfControl.cpp:
```diff
@@ -0,0 +1,772 @@
+                if (node->is<IR::Declaration_Instance>()) {
+                    auto di = node->to<IR::Declaration_Instance>();
+                    cstring name = di->name.name;
+                    auto ctr = new UBPFRegister(program, ctrblk, name, codeGen);
```
Done.

osinstom commented on backends/ubpf/ubpfControl.cpp:
```diff
@@ -0,0 +1,772 @@
+                method->object->getName().name);
+        BUG_CHECK(table != nullptr, "No table for %1%", method->expr);
+
+        P4::ParameterSubstitution binding;
```
Removed.

osinstom commented on backends/ubpf/ubpfControl.cpp:
```diff
@@ -0,0 +1,772 @@
+                                              control->program->refMap,
+                                              control->program->typeMap);
+        auto apply = mi->to<P4::ApplyMethod>();
+        if (apply != nullptr) {
```
Done.

osinstom commented on backends/ubpf/ubpfControl.cpp:
```diff
@@ -0,0 +1,772 @@
+    }
+
+    void UBPFControlBodyTranslator::appendValueAtOperator(const IR::AssignmentStatement *a) const {
+        if (a->right->is<IR::PathExpression>()) {
```
We've refactored the parts related to Registers in ubpfControl. This function has been removed.

osinstom commented on backends/ubpf/ubpfControl.cpp:
```diff
@@ -0,0 +1,772 @@
+    UBPFControlBodyTranslator::preorder(const IR::Operation_Binary *b) {
+        widthCheck(b);
+        builder->append("(");
+        if (b->left->is<IR::PathExpression>()) {
```
That's right. We have refactored it. Now, it is handled in `preorder(IR::PathExpression)`.

osinstom commented on backends/ubpf/ubpfControl.cpp:
```diff
@@ -0,0 +1,772 @@
+
+        if (declType->name.name == p4lib.packetOut.name) {
+            if (method->method->name.name == p4lib.packetOut.emit.name) {
+                ::error("Emit extern not supported in control block");
```
Done.

osinstom commented on backends/ubpf/ubpfControl.cpp:
```diff
@@ -0,0 +1,772 @@
+                    addPadding(paddingInitializers, remainingBits, paddingIndex);
+                    overallPadding -= remainingBits;
+                } else {
+                    error("Not supported bitwidth");
```
This piece of code has been refactored.

osinstom commented on backends/ubpf/ubpfControl.h:
```diff
@@ -0,0 +1,137 @@
+    public:
+        const UBPFControl *control;
+        std::set<const IR::Parameter *> toDereference;
+        std::vector<cstring> saveAction;
```
This is used to implement `action_run`. It stores the action that was executed. This approach is based on and taken from `p4c-ebpf`.

osinstom commented on backends/ubpf/ubpfControl.cpp:
```diff
@@ -0,0 +1,772 @@
+    }
+
+    bool UBPFControlBodyTranslator::preorder(const IR::ReturnStatement *) {
+        builder->appendFormat("goto %s;", control->program->endLabel.c_str());
```
I've added test. This method is applied only for controls. To elimiate returns/exits from actions we use `convertActionBody()`.

osinstom commented:
@mbudiu-vmw I think we addressed most of your comments. There are two actions on our TODO list for next days:
- implement checksum compuation based on p4c-xdp
- refactor packet size's check, as it needs to be more sophisticated. Current approach is too generic as it compares `pkt_len` with the size of the whole `Headers_t` structure.

I'll update you once we will implement this two features.

osinstom commented:
@mbudiu-vmw Could you guide me how to replicate test environment of Travis locally? I'm running tests locally and they are passed, but once I push them to master, Travis-generated tests fail. I use `./bootstrap.sh` script with different options for `UNIFIED` and `GMP`, but environments still seems to be different.

mbudiu-vmw commented:
You have 3 failures in your Travis tests: https://travis-ci.org/p4lang/p4c/jobs/649504069?utm_medium=notification&utm_source=github_status
One of them is a compiler crash, which you should be able to reproduce without any special setup.
The other two are differences in the output files produced by the compiler that you have committed; probably you have to update these files for the tests to pass.

osinstom commented on backends/ubpf/README.md:
```diff
@@ -0,0 +1,87 @@
+### Known limitations
+
+* No support for some P4 constructs (meters, counters, etc.)
+* No support for checksum computation.
```
We've added support for checksums.

osinstom commented:
@mbudiu-vmw Finally, we've made all modifications and tests are passed. I think we can move on with review and merge.

jnfoster commented:
🚀

mbudiu-vmw commented:
Great, I will try to review it next week.

mbudiu-vmw commented on backends/ubpf/examples/vxlan.p4:
```diff
@@ -0,0 +1,224 @@
+            NoAction;
+        }
+
+        //const default_action = NoAction();
```
if you don't need this you should probably delete it

mbudiu-vmw commented on backends/ubpf/target.cpp:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    void UbpfTarget::emitUbpfHelpers(EBPF::CodeBuilder *builder) const {
+        builder->append(
```
this looks a bit brittle. Is there a guarantee that the numbers associated with these functions will never change?

osinstom replied:
This is a strict requirement, which is specific to target that we use (P4rt-OVS). Perhaps, it could be generalized in the future, but at the moment the BPF program would not work otherwise.

mbudiu-vmw commented on backends/ubpf/ubpfDeparser.cpp:
```diff
@@ -0,0 +1,357 @@
+        std::map<const IR::Parameter*, const IR::Parameter*> substitution;
+
+        bool illegal(const IR::Statement* statement)
+        { ::error("%1%: not supported in deparser", statement); return false; }
```
you should consider eventually updating to use an ErrType argument.

osinstom replied:
Done.

mbudiu-vmw commented on backends/ubpf/ubpfTable.cpp:
```diff
@@ -0,0 +1,412 @@
+                    } else if (auto block = s->to<IR::BlockStatement>()) {
+                        convertActionBody(&block->components);
+                        continue;
+                    } else if (s->is<IR::ReturnStatement>()) {
```
Do you want to put here a BUG_CHECK that no return or exit is encountered?

osinstom replied:
I shouldn't put BUG_CHECK here, because retrun or exit could be encountered here (as both exit and return are allowed for actions according to specs). However, we use this `else if` block to implement a proper behavior, i.e. don't emit any statements following exit or return.

osinstom commented on backends/ubpf/examples/packet-counter.p4:
```diff
@@ -0,0 +1,101 @@
+#include <ubpf_model.p4>
+#include <core.p4>
+
+@ethernetaddress typedef bit<48> EthernetAddress;
```
Removed.

osinstom commented on backends/ubpf/README.md:
```diff
@@ -0,0 +1,84 @@
+The sample P4 programs are located in `examples/` directory. We have tested them with the [P4rt-OVS](https://github.com/Orange-OpenSource/p4rt-ovs) switch - 
+the Open vSwitch that can be extended with BPF programs at runtime. See [the detailed tutorial](./docs/EXAMPLES.md) on how to run and test those examples.
+
+The P4 programs for P4-to-uBPF compiler must be written for the `ubpf_model.p4`.
```
Removed.

mbudiu-vmw commented:
Since there are no other comments I will go ahead and merge this huge PR.

jnfoster commented:
Yay!
